### DISCUSSION TITLE
Why is the code below why version 4 cannot access httpserver, but version 3 can?

### DISCUSSION THREAD

**Discussion by liu1998-yowu** [actix/actix-web#3244] [Q&A] [answered]
version 4 code:
```
use std::{sync::mpsc::{Sender, self}, thread::{self, sleep}, time::Duration};
use actix_web::{dev::Server, rt, HttpServer, App, web};

#[actix_web::main]
async fn main() {
    let (tx,rx) = mpsc::channel();
    sapwn_server(&tx).await;
    let _srv = rx.recv().unwrap();
    
    sleep(Duration::from_secs(10));
}
async fn sapwn_server(
    tx:&Sender<Server>
){
    let tx =tx.to_owned();
    thread::spawn(move ||{
        let sys = rt::System::new();
        let pre_srv = HttpServer::new(||{
            App::new()
            .service(web::resource("/").to(|| async { "hello world" }))
        }).bind(("0.0.0.0", 8999));
        let srv = pre_srv.unwrap().run();
        let _ = tx.send(srv);
        sys.run()
    });
}
```
After cargo run, the port is open, and hello world cannot be accessed through the browser.
![image](https://github.com/actix/actix-web/assets/81906221/8130c456-ce1e-4727-ae83-0012ec5585b8)
![image](https://github.com/actix/actix-web/assets/81906221/a14bbb5b-db44-4f8b-b46d-fb0f5e55aac1)

version 3 code:
![image](https://github.com/actix/actix-web/assets/81906221/1b4daf8a-ba26-4a93-b127-bac20124999c)

**Comment by robjtede** [comment, marked as answer]
See this section of the 3->4 migration guide: https://github.com/actix/actix-web/blob/master/actix-web/MIGRATION-4.0.md#server-must-be-polled-warning

**Comment by Samielakkad** [comment]
In Actix Web 4, creating the server future is not enough; the server future must be polled. In your thread, you call:

```rust
let srv = pre_srv.unwrap().run();
let _ = tx.send(srv);
sys.run()
```

But the `Server` you sent through the channel is not being awaited/spawned inside that Actix system. The socket may be bound, but the server is not actually being driven properly.

A simpler pattern is to run/await the server inside the spawned thread:

```rust
thread::spawn(move || {
    actix_web::rt::System::new().block_on(async move {
        HttpServer::new(|| {
            App::new().service(
                web::resource("/").to(|| async { "hello world" })
            )
        })
        .bind(("0.0.0.0", 8999))
        .unwrap()
        .run()
        .await
        .unwrap();
    });
});
```

If you need a handle for stopping the server, use `server.handle()` before awaiting it:

```rust
let server = HttpServer::new(...).bind(...)?.run();
let handle = server.handle();
tx.send(handle).unwrap();
server.await?;
```

So the key migration difference is: in v4, the `Server` future needs to be awaited/spawned; just constructing and sending it around does not make it serve requests.
